Refactor of tropopause_find and dependencies

brigaudet opened the pull request:
1)  Remove code for 'TROP_ALG_ANALYTIC', 'TROP_ALG_STOBIE', and 'TROP_ALG_WMO' tropopause finding options in physics/cam/tropopause.F90, since these options are never invoked externally, and 'output_all' flag is set to .false.
2)  Remove other subroutines and code in tropopause.F90 never invoked.
3)  Extract fields from 'state' structure, and pass these to tropopause_find subroutine, instead of state structure itself.
4)  Clean code (add units, remove single character variable names, insert in / out designations).

Commit:
Initial cleanup of units, terms in tropopause.F90

Commit:
Remove branch for output_all = 'true' option, which would call otherwise unused tropopause options.

Commit:
Remove unused tropopause finding options (TROP_ALG_ANALYTIC, TROP_ALG_STOBIE, TROP_ALG_WMO)

Commit:
Cleanup.

Commit:
1)  Dereference fields from state before tropopause_find is called, and pass in as arguments.
2)  Clean up comments, units, and variable names in twmo and tropopause_cpp.

Commit:
Final cleanup, add 'in / out' notes

jeff-cohere reviewed (approved):
Looks great!

tangshuaiqi commented on components/eam/src/physics/cam/tropopause.F90:
```diff
@@ -476,49 +384,49 @@ subroutine tropopause_climate(pstate, tropLev, tropP, tropT, tropZ)
+        do mn = 11,1,-1
+           if( calday >= days(mn) ) then
               exit
            end if
```
remove the space in "end if" and "end do"?

tangshuaiqi commented on components/eam/src/physics/cam/tropopause.F90:
```diff
@@ -476,49 +384,49 @@ subroutine tropopause_climate(pstate, tropLev, tropP, tropT, tropZ)
+        dels = (calday - days(mn)) / (days(mn+1) - days(mn))
       end if
       
       dels = max( min( 1._r8,dels ),0._r8 )
```
do we want to use min_max_bound here?

brigaudet replied:
Yes, this has been added.

tangshuaiqi commented on components/eam/src/physics/cam/tropopause.F90:
```diff
@@ -722,52 +557,62 @@ end subroutine tropopause_stobie
       ! dt/dz valid?
-      if (dtdz.le.gam) cycle main_loop    ! no, dt/dz < -2 K/km
+      if (dtdz.le.gam)   cycle main_loop    ! no, dt/dz < -2 K/km
       if (pm.gt.plimu)   cycle main_loop    ! no, too low
```
convert .le., .gt. to <= and >

brigaudet replied:
Done.

brigaudet replied:
Also corrected a few other instances of this.

tangshuaiqi commented on components/eam/src/physics/cam/tropopause.F90:
```diff
@@ -788,16 +633,16 @@ subroutine twmo(t, p, plimu, pliml, gam, trp)
+        a2 = (temp1d(jj-1)-temp1d(jj))                     ! a
+        a2 = a2/(pmid1d(jj-1)**cnst_kap-pmid1d(jj)**cnst_kap)
+        b2 = temp1d(jj)-(a2*pmid1d(jj)**cnst_kap)          ! b
         tm2 = a2 * pmk2 + b2                     ! T mean
```
can these separated into a subroutine? Looks like similar code occurs a few times above.

brigaudet replied:
Good idea!  There is now a new subroutine 'get_dtdz'.

tangshuaiqi reviewed (changes requested):
Great work! I just have a few minor comments.

singhbalwinder commented on components/eam/src/physics/cam/tropopause.F90:
```diff
@@ -476,49 +384,49 @@ subroutine tropopause_climate(pstate, tropLev, tropP, tropT, tropZ)
       else
-        do m = 11,1,-1
-           if( calday >= days(m) ) then
+        do mn = 11,1,-1
```
We can store constants like 365 in named constants/parameters, such as:
`integer, parameter :: days_in_year = 365`
`integer, parameter :: months_in_year = 12`

Number `11` can be `months_in_year -1`

brigaudet replied:
This is done.

Commit:
Fix some outstanding formatting issues, create new subroutine get_dtdz for duplicated code.

Commit:
Clean up new code.

singhbalwinder commented on components/eam/src/physics/cam/tropopause.F90:
```diff
@@ -1257,126 +973,131 @@ end subroutine tropopause_findUsing
+        dTdlogP = (temp(icol, tropLev) - temp(icol, tropLev - 1)) / & 
+          (log(pmid(icol, tropLev)) - log(pmid(icol, tropLev - 1)))
+        tropT = temp(icol, tropLev) + (log(tropP) - log(pmid(icol, tropLev))) * dTdlogP
+      endif
```
The calculation above is repeated 4 times (2 times in the tropopause_interpolateZ function). Would it make sense to factor it out?

brigaudet replied:
There are a lot of similar calculations, but enough differences such that I am not sure how much would be saved by factoring out.  The line "tropT = temp(icol, tropLev) + (log(tropP) - log(pmid(icol, tropLev))) * dTdlogP" is repeated twice here, a similar line is repeated twice in interpolateZ but with dZdlogP -- these could be condensed at least.

singhbalwinder replied:
You are right, the operands around the operators can take different places making it hard to refactor. Let's leave it alone. If you have factored out just one line, that is fine too.

singhbalwinder commented on components/eam/src/physics/cam/tropopause.F90:
```diff
@@ -1257,126 +973,131 @@ end subroutine tropopause_findUsing
   ! This routine interpolates the temperatures in the physics state to
   ! find the temperature at the specified tropopause pressure.
-  function tropopause_interpolateT(pstate, icol, tropLev, tropP)
+  function tropopause_interpolateT(pmid,temp,icol,tropLev,tropP)
```
Since this and the `tropopause_interpolateZ` functions do not have loops, we probably should send single values rather than the whole array. For example, for this function, we should send only pmid(icol, tropLev), pmid(icol, tropLev+1), pmid(icol, tropLev-1), etc. Sending `icol` and `tropLev` may not be required. This will make the code easier to understand and get rid of a lot of array lookups.

brigaudet replied:
This is done, thanks.

singhbalwinder commented on components/eam/src/physics/cam/tropopause.F90:
```diff
@@ -1257,126 +973,131 @@ end subroutine tropopause_findUsing
   ! This routine interpolates the pressures in the physics state to
   ! find the pressure at the specified tropopause altitude.
-  function tropopause_interpolateP(pstate, icol, tropLev, tropZ)
+  function tropopause_interpolateP(pmid,zm,icol,tropLev,tropZ)
```
I wrote a comment below about using explicit single values rather than sending arrays. It applies to this function as well. ( Sorry, I am reviewing it backward :smile:)

brigaudet replied:
See above.

singhbalwinder commented on components/eam/src/physics/cam/tropopause.F90:
```diff
@@ -816,64 +649,89 @@ subroutine twmo(t, p, plimu, pliml, gam, trp)
+    real(r8), intent(in)                    :: temp1d_up     !  temperature in column at upper level [K]
+    real(r8), intent(in)                    :: temp1d_down   !  temperature in column at lower level [K]
+    real(r8), intent(inout)                 :: dtdz     ! temperature lapse rate vs. height [K/m]
+    real(r8), intent(inout)                 :: tm       ! mean temperature [K] -- needed to find pressure at trop + 2 km
```
`dtdz` and `tm` can be `out` instead of `inout`

brigaudet replied:
Thanks, done.

singhbalwinder commented on components/eam/src/physics/cam/tropopause.F90:
```diff
@@ -568,39 +481,38 @@ subroutine tropopause_hybridstobie(pstate, tropLev, tropP, tropT, tropZ)
     !    write(iulog,*) 'In set_chem_trop, o3_ndx =',o3_ndx
```
Remove the commented-out code.

brigaudet replied:
OK, done.

singhbalwinder commented on components/eam/src/physics/cam/tropopause.F90:
```diff
@@ -882,218 +740,123 @@ subroutine tropopause_twmo(pstate, tropLev, tropP, tropT, tropZ)
       ! Skip column in which the tropopause has already been found.
-      if (tropLev(i) == NOTFOUND) then
+      if (tropLev(icol) == NOTFOUND) then
         tmin = 1e6_r8
```
This should actually be set to `huge(tmin)` but we probably can't change that now.

brigaudet replied:
Added a comment to note this.

singhbalwinder commented on components/eam/src/physics/cam/tropopause.F90:
```diff
@@ -568,39 +481,38 @@ subroutine tropopause_hybridstobie(pstate, tropLev, tropP, tropT, tropZ)
-       not_found: if (tropLev(i) == NOTFOUND) then
+       not_found: if (tropLev(icol) == NOTFOUND) then
 
           stobie_min = 1.e10_r8    ! An impossibly large number
```
This should be set to `huge(stobie_min)` but we probably can't change it now.

brigaudet replied:
Added a comment to note this.

singhbalwinder commented on components/eam/src/physics/cam/tropopause.F90:
```diff
@@ -476,49 +385,48 @@ subroutine tropopause_climate(pstate, tropLev, tropP, tropT, tropZ)
-          tP = tropp_p_loc(i,lchnk,last) &
-            + dels * (tropp_p_loc(i,lchnk,next) - tropp_p_loc(i,lchnk,last))
+          tP = tropp_p_loc(icol,lchnk,last) &
+            + dels * (tropp_p_loc(icol,lchnk,next) - tropp_p_loc(icol,lchnk,last))
```
This part of the code uses the data (read from a file) based on the location of `icol` (or column) on the globe. This needs to be modified based on how we read files on the EAMxx side. Let's leave a note here (e.g., "`tropp_p_loc` has location-based data as a function of `icol` and `lchnk`, C++ ported code needs to read in the data and use the data based on the column it is operating on") and we need to address it in the ported C++ code.

brigaudet replied:
OK, done, thanks.

singhbalwinder reviewed (changes requested):
Thanks @brigaudet for refactoring this complicated routine. It looks really nice! I have left some minor comments and asked for some changes. Please let me know if you have any questions.

Commit:
1)  Change arguments of interpolation functions to be single values, not arrays.
2)  Add calendar-related parameters.
3)  Add refactoring comments.
4)  Some more code cleanup.
5)  Fix incorrect actual argument sequencing in function call.

singhbalwinder reviewed (approved):
Thanks for addressing all the comments. This is ready to go now!

Commit:
1)  Correct sequencing of pint, pmid in calls to tropopause_find in tropopause_output.
2)  Fix type of days_in_year_real parameter

brigaudet commented:
Longer (1-year) tests were failing BFB, most recent commit fixed this by correcting sequence of some argument lists in tropopause_output.  Also fixed an incorrect type declaration of a parameter.  Most recent commit passes all tests:
![image](https://github.com/eagles-project/e3sm_mam4_refactor/assets/48535869/94f58274-45f7-4ed8-a6bc-aba8e25bcd83)
